File: 8-1.java
import java.util.*;

public class Main {
    public static List<String> split(String s, char delimiter) {
        List<String> tokens = new ArrayList<>();
        StringBuilder token = new StringBuilder();
        for (char c : s.toCharArray()) {
            if (c == delimiter) {
                tokens.add(token.toString());
                token = new StringBuilder();
            } else {
                token.append(c);
            }
        }
        tokens.add(token.toString());
        return tokens;
    }

    public static Pair<Integer, String> parse_log(String s) {
        List<String> tokens = split(s, ' ');
        int t = Integer.parseInt(tokens.get(0).substring(0, 2)) * 60 + Integer.parseInt(tokens.get(0).substring(3, 5));
        String name = tokens.get(1);
        return new Pair<>(t, name);
    }

    public static int get_fee(List<Integer> fees, int t) {
        int money = fees.get(1);
        if (fees.get(0) < t) {
            money += (t - fees.get(0) + fees.get(2) - 1) / fees.get(2) * fees.get(3);
        }
        return money;
    }

    public static List<Pair<String, Integer>> solution(int n, List<String> A, List<Integer> fees) {
        Map<String, Integer> d = new HashMap<>();
        for (String log : A) {
            Pair<Integer, String> parsed_log = parse_log(log);
            int t = parsed_log.getFirst();
            String name = parsed_log.getSecond();
            if (d.containsKey(name)) {
                d.put(name, d.get(name) + t);
            } else {
                d.put(name, t);
            }
        }
        List<Pair<String, Integer>> answer = new ArrayList<>();
        for (Map.Entry<String, Integer> entry : d.entrySet()) {
            entry.setValue(get_fee(fees, entry.getValue()));
            answer.add(new Pair(entry.getKey(), entry.getValue()));
        }
        
        answer.sort((a, b) -> {
            if (!a.getSecond().equals(b.getSecond())) {
                return b.getSecond() - a.getSecond();
            } else {
                return a.getFirst().compareTo(b.getFirst());
            }
        });
        return answer;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        scanner.nextLine();
        List<String> A = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            A.add(scanner.nextLine());
        }
        List<Integer> fees = Arrays.asList(100, 10, 50, 3);
        List<Pair<String, Integer>> B = solution(n, A, fees);
        for (Pair<String, Integer> entry : B) {
            System.out.println(entry.getFirst() + " " + entry.getSecond());
        }
    }
}

class Pair<T, U> {
    private T first;
    private U second;

    public Pair(T first, U second) {
        this.first = first;
        this.second = second;
    }

    public T getFirst() {
        return first;
    }

    public U getSecond() {
        return second;
    }
}

File: 8-1.cpp
#include <bits/stdc++.h>
using namespace std;

vector<string> split(string s, char delimiter) {
    vector<string> tokens;
    string token;
    for (char c : s) {
        if (c == delimiter) {
            tokens.push_back(token);
            token = "";
        }
        else {
            token += c;
        }
    }
    tokens.push_back(token);
    return tokens;
}

pair<int, string> parse_log(string s) {
    vector<string> tokens = split(s, ' ');
    int t = stoi(tokens[0].substr(0, 2)) * 60 + stoi(tokens[0].substr(3, 2));
    string name = tokens[1];
    return make_pair(t, name);
}

int get_fee(vector<int> fees, int t) {
    int money = fees[1];
    if (fees[0] < t) {
        money += (t - fees[0] + fees[2] - 1) / fees[2] * fees[3];
    }
    return money;
}

vector<pair<string, int>> solution(int n, vector<string> A, vector<int> fees) {
    map<string, int> d;
    for (auto log : A) {
        pair<int, string> parsed_log = parse_log(log);
        int t = parsed_log.first;
        string name = parsed_log.second;
        if (d.find(name) != d.end()) {
            d[name] += t;
        }
        else {
            d[name] = t;
        }
    }
    for (auto& entry : d) {
        entry.second = get_fee(fees, entry.second);
    }
    vector<pair<string, int>> answer(d.begin(), d.end());
    sort(answer.begin(), answer.end(), [](const pair<string, int>& a, const pair<string, int>& b) {
        if (a.second != b.second) {
            return a.second > b.second;
        }
        else {
            return a.first < b.first;
        }
        });
    return answer;
}

int main() {
    int n;
    scanf("%d\n", &n);
    vector<string> A(n);
    for (int i = 0; i < n; i++) {
        getline(cin, A[i]);
    }
    vector<int> fees = { 100, 10, 50, 3 };
    vector<pair<string, int>> B = solution(n, A, fees);
    for (auto entry : B) {
        cout << entry.first << " " << entry.second << endl;
    }
    return 0;
}
